File: code/scenes/src/main/java/org/fbme/scenes/controllers/components/ComponentEntry.java
package org.fbme.scenes.controllers.components;

import org.fbme.scenes.controllers.LayoutSetting;

import java.awt.*;
import java.util.function.Supplier;

class ComponentEntry<CompT, FormT> {
    final CompT myComponent;

    final MyLayoutSetting myLayoutSetting;
    Supplier<FormT> myFormProvider;
    FormT myModelForm;
    FormT myTransformedForm;

    final ComponentController<FormT> myController;

    private final ComponentsFacility<CompT, FormT> myFacility;

    ComponentEntry(ComponentsFacility<CompT, FormT> facility, CompT component) {
        myComponent = component;
        myLayoutSetting = new MyLayoutSetting();
        myFacility = facility;
        myController = myFacility.myControllerFactory.create(myFacility.myEditor.getContext(), myComponent);
        myFormProvider = myFacility.myComponentSyncronizer.getForm(myComponent);
        myModelForm = myFormProvider.get();
        myLayoutSetting.myBounds = myController.getBounds(myModelForm);
    }

    void relayout() {
        myModelForm = myFormProvider.get();
        myLayoutSetting.myBounds = myController.getBounds(myModelForm);
        myController.updateCellWithForm((myTransformedForm != null ? myTransformedForm : myModelForm));
        myController.updateCellSelection(isSelected());
    }

    Rectangle getLayoutBounds() {
        Rectangle bounds = myController.getBounds(myModelForm);
        if (myTransformedForm != null) {
            bounds = bounds.union(myController.getBounds(myTransformedForm));
        }
        return bounds;
    }

    public boolean isSelected() {
        return myFacility.mySelection.isSelected(myComponent);
    }

    private class MyLayoutSetting implements LayoutSetting {
        private Rectangle myBounds;

        @Override
        public Rectangle getBounds() {
            return myBounds;
        }

        @Override
        public boolean canStartMoveAt(int x, int y) {
            return myController.canStartMoveAt(myModelForm, x, y);
        }

        @Override
        public void moveTo(int x, int y) {
            myFacility.myComponentSyncronizer.setForm(myComponent, myController.translateForm(myModelForm, x - myBounds.x, y - myBounds.y));
        }
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/components/ComponentEntry.kt
package org.fbme.scenes.controllers.components

import org.fbme.scenes.controllers.LayoutSetting
import java.awt.Rectangle

class ComponentEntry<CompT, FormT>(facility: ComponentsFacility<CompT, FormT>, val myComponent: CompT) {
    val layoutSetting: MyLayoutSetting
    private var formProvider: () -> FormT
    var modelForm: FormT
    var transformedForm: FormT? = null
    val controller: ComponentController<FormT>
    private val facility: ComponentsFacility<CompT, FormT>

    fun relayout() {
        modelForm = formProvider()
        layoutSetting.bounds = controller.getBounds(modelForm)
        controller.updateCellWithForm(transformedForm ?: modelForm)
        controller.updateCellSelection(isSelected)
    }

    val layoutBounds: Rectangle
        get() {
            val bounds = controller.getBounds(modelForm)
            return if (transformedForm == null) bounds else bounds.union(controller.getBounds(transformedForm!!))
        }

    private val isSelected: Boolean
        get() = facility.selection.isSelected(myComponent)

    inner class MyLayoutSetting : LayoutSetting {
        override var bounds: Rectangle = Rectangle()

        override fun canStartMoveAt(x: Int, y: Int): Boolean {
            return controller.canStartMoveAt(modelForm, x, y)
        }

        override fun moveTo(x: Int, y: Int) {
            facility.componentSynchronizer.setForm(
                myComponent,
                controller.translateForm(modelForm, x - bounds.x, y - bounds.y)
            )
        }
    }

    init {
        layoutSetting = MyLayoutSetting()
        this.facility = facility
        controller = this.facility.controllerFactory.create(this.facility.editor.editorContext, myComponent)
            ?: error("Can't get controller")
        formProvider = this.facility.componentSynchronizer.getForm(myComponent)
        modelForm = formProvider()
        layoutSetting.bounds = controller.getBounds(modelForm)
    }
}
